Mup 1.5

zodern opened the pull request:
Mup 1.5 adds load balancing, zero downtime deploys, performance improvements, and can optionally use Docker Swarm.

### Current Status (March2020)

**Load balancing and Zero Downtime Deploys** Mup will be able to optionally load balance across multiple servers with sticky sessions. When this is enabled and more than 1 server is used, deploys will have no downtime. It uses nginx and the reverse proxy for this. The mup config now has a `privateIp` option. When it is configured, the requests between nginx and other servers will go over the private network.

**Prepare Bundle Performance**  For apps with a large bundle, the `Prepare Bundle` step could be very slow. When the `app.docker.useBuildKit` option is enabled, mup will use Docker's new BuildKit image builder and one of its experimental feature to reduce the time spent building the image by 60%.

**Private Docker Registry** Deploying an app to many servers was slow since Mup would upload the app's bundle to each server, and then either build a docker image (if using prepare bundle) or install the npm dependencies on each server. Mup can now be configured to use a private docker registry. It will then only upload the bundle to a single server, run `Prepare Bundle` to create a docker image, and store the image in the registry. For the rest of the servers it will use the image from the registry instead of building a new one.

**Production Debugging** Run `mup meteor debug` to connect your local node debugger (such as Chrome's node inspector) to your app running on the server

**Docker Swarm** Mup is able to setup and manage a swarm cluster. When using swarm, load balancing and zero downtime deploys are always enabled. There are a couple situations it doesn't correctly handle yet, listed in #893.

Most of the reasons why I created the swarm integration no longer exist, and there is a new simpler implementation for load balancing and zero downtime deploys. For now I plan to leave the feature in and see if there is any interest in the feature or anyone willing to maintain it. When I start work on mup 2 (probably 8+ months away) I will re-evaluate and see if there was any interest or anyone willing to maintain it.

**Docker Services** When swarm is enabled, the app is deployed as a swarm service. There are a couple of commands that haven't been updated yet. In the final release, there will be some config options not supported, though I will try to add support in future releases. It also doesn't have a good experience around failed deploys.

**Load Balancing with Docker Swarm** Uses [nginx-proxy-swarm-upstream](https://github.com/zodern/nginx-proxy-swarm-upstream). It works, and sticky sessions are enabled.

**Changing default docker image** The current default docker image only supports Meteor 1.3 and older, so most users need to modify the config to use the correct docker image. I have been working on the [`zodern/meteor`](https://github.com/zodern/meteor-docker) docker image which supports Meteor 1.2 and newer with one image. To avoid breaking backwards compability, it could default to using the current default image for Meteor 1.2 - 1.3, and `zodern/meteor` for newer versions of Meteor.

**Performance** The Start Meteor task list will finish 12 seconds sooner when Prepare Bundle is enabled. If Mongo is already running, and its config hasn't changed, the `Mongo Start` task list will only be a couple of seconds instead of around 30. Also, ssh sessions are kept connected between task lists instead of disconnecting and reconnecting. Mup's tests finish over 15 minutes faster with mup 1.5. Some task lists for the proxy are run no longer run in serial when using multiple servers.

#### Beta

You can try the beta with `npm i -g mup@next`. You can view the new docs [here](https://github.com/zodern/meteor-up/pull/975/files?short_path=c331e99#diff-c331e99b809715805fe7447af020d198), and the list of changes in the [changelog](https://github.com/zodern/meteor-up/pull/975/files?short_path=c331e99#diff-4ac32a78649ca5bdd8e0ba38b7006a1e).

zodern committed:
Merge branch 'master' into dev

zodern committed:
Merge branch 'dev' into origin/mup-1.5

zodern committed:
Run meteor app as swarm service

zodern committed:
Run meteor app as swarm service

zodern committed:
Simplify swarm api's

- Use server's name instead of host in more places
- Labels returned by swarmOptions is an array instead of an object

zodern committed:
Simplify swarm api's

- Use server's name instead of host in more places
- Labels returned by swarmOptions is an array instead of an object

zodern committed:
Store validation errors even when they aren't shown

zodern committed:
Add noDefaults to VALIDATE_OPTIONS

zodern committed:
Fix starting app

zodern committed:
Update eslint config to warn if no new line after var

zodern committed:
Add api.tasks with tasks for docker swarm

zodern committed:
Allow using nodemiral session in runSSHCommand

zodern committed:
Add pluginApi.dockerServiceInfo

zodern committed:
Use addCreateOrUpdateService for starting the Meteor app

zodern committed:
Increase concurrency of serverInfo to 3

zodern committed:
Update eslint config

zodern committed:
Merge remote-tracking branch 'origin/mup-1.5' into mup-1.5

zodern committed:
Re-use sessions between task lists and runSSHCommand

zodern committed:
Add stop service task

zodern committed:
Add load balancing

zodern committed:
Disable checking for updates in tests

zodern committed:
Fix running tasks on managers when setting up swarm

zodern committed:
Update Meteor app used in tests to 1.7
A number of packages were also removed to help speed up tests.

zodern committed:
Fix proxy status when proxy.servers in the config is used

zodern committed:
Show better message when unable to add or remove a node label

zodern committed:
Fix some managers not being promoted after initializing swarm

zodern committed:
Remove unused props and exports in swarm utils

zodern committed:
Throw error when servers are in multiple clusters
Any servers already in a cluster need to be in the same cluster. This tries to show helpful information so the user can decide how to fix it.

I had considered having mup try to automatically fix this, but that could easily cause other clusters or apps to stop working.

zodern committed:
Improve task name when destroying swarm cluster

zodern committed:
docker status shows "Unknown server" for nodes not in the config

zodern committed:
Safely demote nodes

zodern committed:
Use smaller image in tests to fix travis no space on disk error

zodern committed:
Node being joined should not already be part of cluster

zodern committed:
Validate swarm object in config

zodern committed:
Document load balancing and swarm

zodern committed:
Add 'nan' to list of packages to remove from shrinkwrap

zodern committed:
Fix lint error

zodern committed:
Document minimum memory servers should have when using Swarm

zodern committed:
Fix syslog test

zodern committed:
app.docker.imagePort defaults to 3000 instead of 80All of the docker images I know of that are compatible with Meteor Up allow using the PORT environment variable to change what port the app uses, so this change should not break compatibility.

zodern committed:
Fix docker ps
Since sessions are now closed after the last promised is resolved, commands and hooks using sessions need to return a promise.

zodern committed:
Remove unused swarmServices from server info

zodern committed:
Remove unused task

zodern committed:
Wait less time after starting the app and prepare bundle is enabled

zodern committed:
Only restart mongodb container when necessary
The container is started when it isn't running, or the start script has changed.

zodern committed:
Set NODE_VERSION build arg during prepare bundle

zodern committed:
Combine ENV build instructions in Prepare Bundle
This saved ~30 seconds in one of mup's tests

zodern committed:
Fix verifying deployment with non-root user

zodern committed:
Use title case for stop mongo task name

zodern committed:
Update changelog

zodern committed:
Revert "Update Meteor app used in tests to 1.7 A number of packages were also removed to help speed up tests."

This reverts commit 27b81807795b035d9c62fb8f47472fea14d2fd51.
The Meteor 1.7 app takes over 90 seconds longer to build for the tests.

zodern committed:
Remove newline after node version from .node_version.txt

zodern committed:
Fix load balancing services listening on ports other than 80

zodern committed:
Update changelog

zodern committed:
Merge branch 'master' into mup-1.5

zodern committed:
Merge branch 'mup-1.5' into dev

zodern committed:
Release 1.5.0-beta1

zodern committed:
Fix getting docker sessions when swarm is enabled

zodern committed:
Fix docs on sudo

Floriferous commented:
This is an absolutely insane release @zodern, everything is coming together, super excited!

zodern committed:
Support `app.docker.networks` when using Docker Swarm

zodern committed:
Remove unecessary TODO comment

zodern committed:
Fix removing environment variables

zodern committed:
Update meteor plugin's commands to support swarm

zodern committed:
Add empty line before "Collecting Docker information"

zodern committed:
Update service endpoint mode

zodern committed:
Validate for at least one server in `app.servers`

zodern committed:
Only connect app to proxy's network when using swarm

zodern committed:
Fix getting sessions when using swarm in Meteor plugin

zodern committed:
Fix getting sessions when stopping proxy

zodern committed:
Set VIRTUAL_PORT when using reverse proxy

zodern committed:
Add instruction for setting Cloudflare's SSL setting

zodern committed:
Fix viewing service logs

zodern committed:
Improve error thrown by dockerServiceInfo when there are no managers

zodern committed:
Set service's update-failure-action and update-parallelism
The service will rollback when an update fails.
It will update 1/3 of the instances at a time, rounded up.

zodern committed:
Use api.swarmEnabled

zodern committed:
Log message after running zodern/nginx-proxy-swarm-upstream

ninjaPixel committed:
Add new docker image to the 'Meteor Support' table

Meteor 1.8 uses Node 8.11.2

ninjaPixel committed:
Update mup.js.sample with an image that works with Meteor 1.8

ninjaPixel committed:
Add instructions for how to build and serve the docs website

ninjaPixel committed:
Fix lint error

Line was too long

zodern committed:
Fix validating swarm options

zodern committed:
Merge pull request #1024 from ninjaPixel/update-docs-for-meteor-1.8

Update docs for meteor 1.8

zodern committed:
Merge remote-tracking branch 'origin/dev' into dev

zodern committed:
Use ADD instead of COPY in prepare bundle to improve performance

zodern committed:
Add solution for error joining nodes  due to servers having different time

zodern committed:
Fix prepare bundle

zodern committed:
Fix elapsed time in Verifying Deployment

zodern committed:
Increase curl timeout

zodern committed:
Show "=> Logs:" in stderr next to logs

zodern committed:
Fix removing  environment variables with "mup reconfig"
Since Prepare Bundle had added the environment variables  using ENV instructions in the docker file, the only way to remove them was to redeploy.

zodern committed:
Fix using prepare bundle with non-root images

rolljee committed:
Automaticaly goes from http to https when using ssl

rolljee committed:
Fix #784  when unencoded special characters in MONGO_URL and MAIL_URL

rolljee committed:
Fix issue with MAIL_URL not defined

rolljee committed:
Reworked MONGO_URL & MAIL_URL to be encoded the write way

rolljee committed:
Removed some unused var & comment

rolljee committed:
Fixing Cannot read property split of undefined

rolljee committed:
mup docker status should show warning when servers do not have the same docker version.

rolljee committed:
Switch to https://github.com/opencollective/opencollective-postinstall

rolljee committed:
Audit npm deps & update yargs

zodern committed:
Fix running tests with latest version of Docker
The latest version of Docker no longer sets up a service when installed in the docker container.

rolljee committed:
Merge branch 'dev' of https://github.com/zodern/meteor-up into dev

rolljee committed:
Requested changes by Zodern

rolljee committed:
Added false if description is empty

zodern committed:
Small fixes

zodern committed:
Remove console.log

zodern committed:
Switch back to using COPY when building docker image
The performance difference with ADD is small in newer versions of docker, and ADD created many permission problems that are difficult to solve in a way that is compatible with all docker images.

zodern committed:
Update changelog

rolljee committed:
Modified prepareConfig from meteor plugin

rolljee committed:
Fix typo

zodern committed:
Do not wait for update check to finish

eluck committed:
Fix deployment check for multiple networks on host

eluck committed:
Run integration tests again

eluck committed:
Merge branch 'dev' into fix/deployment-check

zodern committed:
Add 20 second update delay

zodern committed:
Document docker volumes

zodern committed:
Merge pull request #1055 from eluck/fix/deployment-check

Fix deployment check for multiple networks on host

zodern committed:
Merge pull request #1025 from ninjaPixel/update-docs-with-how-to-build-docs

Add instructions for how to build and serve the docs website

zodern committed:
Add "mup docker update" command

zodern committed:
Use numbers for app image tag with docker swarm
This is needed for swarm to switch to the old version of the app when a deploy fails.

zodern committed:
Only run service on servers listed in app.servers

zodern committed:
Handle missing properties when in meteor status

zodern committed:
Use supplied tag in prepare-bundle

zodern committed:
Fix meteor swarm test

zodern committed:
Merge branch 'dev' into dev

zodern committed:
Merge pull request #1051 from rolljee/dev

Modified prepareConfig from meteor plugin

zodern committed:
Fix lint error

zodern committed:
Show warning when using swarm and not proxy

zodern committed:
Merge branch 'master' into dev

zodern committed:
Reorganize changes for 1.5

zodern committed:
Release 1.5.0-beta.2

zodern committed:
Support multiple validators for same root path

zodern committed:
Update ssh2 and nodemiral

zodern committed:
Increase server info concurrency

zodern committed:
Do not remove labels users added

zodern committed:
Add swarm.labels option

zodern committed:
Add proxy.shared.nginxTemplate option

zodern committed:
Fix proxy for apps not running as a swarm service

zodern committed:
Release 1.5.0-beta.3

zodern committed:
Fix error when shared.sharedNginxConfig is empty

zodern committed:
Always add proxy env vars

zodern committed:
Release 1.5.0-beta.4

zodern committed:
Add stderr to output of runSSHCommand

zodern committed:
Fix stoping proxy

zodern committed:
Validate proxy.shared.templatePath option

zodern committed:
Clean up upstream when app is no longer a service

zodern committed:
Require node 8

zodern committed:
Update dependencies

zodern committed:
Add mongo.dbName option
The reason it was difficult to use the same database with multiple apps is mup would use the app's name as the database name. With this option it is possible to override that.

zodern committed:
Merge branch 'dev' of github.com:zodern/meteor-up into dev

zodern committed:
Fix build npm script

zodern committed:
Fix --servers option

zodern committed:
Support --servers option with "mup ssh"

zodern committed:
Fix using proxy with imagePorts other than 80

zodern committed:
Add load balancing without swarm
This requires the ip address and port for each instance to be predictable. IP addresses are the server's host name. Unlike load balancing with swarm, this does make the container available without going through the proxy. It might be fixable by using a firewall to restrict access to the servers with the proxy.

For the port, we default it to a random port between 10,000 and 20,000, using the app name as a seed. In the future we could run a script on each server first to find an available port if we encounter too many conflicts.

zodern committed:
Add --overview option for status

zodern committed:
Only configure upstream when load balancing

zodern committed:
Add StatusDisplay class
This class takes care of displaying status information and coloring.
It also handles hiding information in the overview mode

zodern committed:
Merge branch 'dev' of github.com:zodern/meteor-up into dev

zodern committed:
Clean up mongo status

zodern committed:
Remove unused code

zodern committed:
Fix "meteor status" when docker shows warnings

zodern committed:
Prevent bash from modifying custom nginx configs

zodern committed:
Use private ip if available for load balancing

zodern committed:
Merge branch 'dev' of https://github.com/zodern/meteor-up into dev

zodern committed:
Add `mup meteor destroy` command to remove app

zodern committed:
Update changelog

zodern committed:
Remove optional deps from shrinkwrap

zodern committed:
Release 1.5.0-beta.6

zodern committed:
Update tar

zodern committed:
Optionally use buildkit and experimental mount
For an app with a 500mb bundle, this reduces building the image from 3.5 minutes to 1.4 minutes.
This implementation gives us more control over how the tar file is extracted, which allows us to avoid the problems with using ADD.

zodern committed:
Merge branch 'dev' of github.com:zodern/meteor-up into dev

zodern committed:
Optionally store image on private docker registry

zodern committed:
Support private docker registries

zodern committed:
Fix PORT warning when using load balancing

zodern committed:
Start proxy in series again

zodern committed:
Update node versions used in travis

zodern committed:
Update babel

zodern committed:
Change how docker is installed for tests

zodern committed:
Release 1.5.0-beta.7

zodern committed:
Update getting started

zodern committed:
Update changelog

zodern committed:
Update docs

zodern committed:
Fix invalid js in docs

zodern committed:
Fix lint errors

zodern committed:
Fix parsing json output when it includes warnings

zodern committed:
Improve checking if app is accessible

zodern committed:
Fix load balancing when app is stopped on a server
The nginx config only includes apps running on the same server. If the app is stopped on the server running nginx, nginx returns an error status code instead of sending requests to other servers.
It now runs a lightweight configuration container to configure the proxy like is done when using Docker Swarm.

zodern committed:
Fix cleaning up upstream configs
It removes upstream configs that are broken symlinks, but these symlinks are always broken outside of the nginx container so it removed valid upstream configs for other apps.

zodern committed:
Merge branch 'dev' of github.com:zodern/meteor-up into dev

zodern committed:
Fix running prepare-bundle concurrently
When multiple apps are being deployed at the same time, the Prepare Bundle script could fail since `docker image prune` fails if it is already running.

zodern committed:
Merge branch 'dev' of https://github.com/zodern/meteor-up into dev

zodern committed:
Release 1.5.0-beta.8

zodern committed:
Add production debugging of Meteor server

zodern committed:
Clean up images when using private registry

zodern committed:
Update changelog

zodern committed:
Release 1.5.0-beta.9

zodern committed:
Fix starting container needed for debugging

zodern committed:
Log instructions when using debugger

zodern committed:
Prune build cache when using build kit

zodern committed:
Release 1.5.0-beta.10

zodern committed:
Fix service-configure script for non-root users

zodern committed:
Fix additional docker command in service-configure

zodern committed:
Fix uploading certificates before app was setup

zodern committed:
Create proxy volume for htpasswd

zodern committed:
Remove autofix for special characters in env vars
Mup would try to automatically url encode special characters in passwords in MONGO_URL and MAIL_URL. For some MONGO_URLS it would modify other parts of the URL also which would break connecting to mongo.

zodern committed:
Add option to disable sticky sessions

zodern committed:
Update changelog

zodern committed:
Release 1.5.0-beta.11

zodern committed:
Merge branch 'master' into dev

zodern committed:
Add new website

zodern committed:
Replace npm-shrinkwrap with package-lock
There once was a patch update to a dependency which broke mup if it was installed after that update. We started using npm-shrinkwrap to avoid that happening again.

New versions of NPM have started including dev and optional dependencies in npm-shrinkwrap, and always installing them when a package is installed globally. We had a script to fix the optional dependencies, but it will be more difficult to remove dev dependencies. For now we will switch back to using package-lock.json.

zodern committed:
Fix MUP_SKIP_UPDATE_CHECK env var

zodern committed:
Fix starting docker for tests
In the image created from "docker commit", docker would fail to start. This switches to using a volume instead to cache docker images when running tests locally.

zodern committed:
Fix rerunning tests locally

zodern committed:
Fix test for deploying Meteor 1.2

zodern committed:
Add additional deploy tests

zodern committed:
Fix tests

zodern committed:
Re-enable all tests

zodern committed:
Reduce logs from building test images

zodern committed:
Update dependencies

zodern committed:
Fix starting app when prepare bundle is disabled

zodern committed:
Run tests with node 14

zodern committed:
Update .gitignore

zodern committed:
Prefer IP Address on bridge network

zodern committed:
Release 1.5.0-beta.12

zodern committed:
Remove silent-npm-registry-client

zodern committed:
Fix comparing beta versions

zodern committed:
Remove babel-polyfill

zodern committed:
Release 1.5.0-beta.13

zodern committed:
Change prepublish script to prepublishOnly

zodern committed:
Update joi

zodern committed:
Change prepublishOnly script to prepare

zodern committed:
Update stylelint

zodern committed:
Fix css lint errors

zodern committed:
Increase margin below image

zodern committed:
Merge branch 'dev' into feature/new-website

zodern committed:
Merge pull request #1151 from zodern/feature/new-website

New Website

zodern committed:
Release 1.5.0-beta.14

zodern committed:
Update homepage content

zodern committed:
Require docker 18 or newer

tjramage commented:
Amazing! Great stuff @zodern 👍

krizka commented:
@zodern Server debugging is mind-blowing feature. Thank you!
